### ISSUE TITLE
[Q]: Out of order logging: Steps must be monotonically increasing, so this data will be ignored

### ISSUE THREAD

**Issue by wdykas** [wandb/wandb#10628]
### Ask your question

Hi I am getting the following warning in my logging code when trying to log things with different types of steps. This is causing me not to log most of my data.

```
WARNING Tried to log to step 0 that is less than the current step 8073. Steps must be monotonically increasing, so this data will be ignored. See https://wandb.me/define-metric to log data out of order.
```

I am attempting to add inference logging code to an RL loop where there is existing training logging to wandb. The problem is that I am logging every inference step which is different than a training step. So I log fine during inference for x amount of inference steps but then the first training logging fails with the above because it is still on step=0. I attempted to solve this by defining a custom step_metric like the following:

```
metrics_writer.define_metric("inference/*", step_metric="inference/inference_step")
metrics_writer.log(metrics, commit=True)
```

While the training loop keeps its regular logging like 

```
metrics_writer.log({'samples vs steps': args.consumed_train_samples}, iteration)
```

However I am still getting this error, is there something else I can do to make this work?

**Comment by abhi-neelam** [comment]
Hey, I had a similar issue. for step metrics, I think you have to add the step_metric as part of the dictionary directly instead of the second parameter of the log function.

Something like this could work.
```
metrics_writer.log({'samples vs steps': args.consumed_train_samples, 'training/training_step': iteration})
```

Please let me know if this helps.

**Comment by Anshuman-WandB** [comment]
Hi @wdykas,

Just wanted to jump in and confirm that the solution provided by @abhi-neelam is correct. Thanks for the great suggestion!

To solve the "monotonically increasing step" you need to do two things:

- Define the relationship: Use wandb.define_metric() to tell the UI which metric to use as the custom x-axis for others.
- Log the custom step: When you call wandb.log(), you must include the custom x-axis metric and its current value within the same dictionary as the other metrics. 

Please refer these docs for more [information](https://docs.wandb.ai/guides/track/log/customize-logging-axes/) and this colab [notebook](https://colab.research.google.com/github/wandb/examples/blob/master/colabs/wandb-log/Customize_metric_logging_with_define_metric.ipynb). 

A Note on Hiding Your Step Metric:
You'll notice that your custom step metrics (e.g., inference/inference_step) will also appear as plots in your workspace, usually creating a straight diagonal line. To prevent this, you can use the `hidden=True` parameter:
```
wandb.define_metric("inference/inference_step", hidden=True)
```
However, please be aware that there is a known bug where hidden=True is not currently working as expected. This means the step metric will still be plotted for now, even if you set it to hidden. Our team is aware of this and tracking a fix.

**Comment by Anshuman-WandB** [comment]
Hi @wdykas,

Following up on this thread regarding the hidden=True parameter for step metrics. Just wanted to confirm that this functionality is working as expected now, and metrics defined with hidden=True (like inference/inference_step in the example) are successfully hidden from the workspace plots.

Since the bug is resolved, I will go ahead and close this issue.
